subtype/src/main/java/com/fasterxml/jackson/module/subtype/SubtypeModule.java

```diff
@@ -0,0 +1,120 @@
+package com.fasterxml.jackson.module.subtype;
+
+import com.fasterxml.jackson.core.Version;
+import com.fasterxml.jackson.databind.AnnotationIntrospector;
+import com.fasterxml.jackson.databind.Module;
+import com.fasterxml.jackson.databind.introspect.Annotated;
+import com.fasterxml.jackson.databind.jsontype.NamedType;
+import com.fasterxml.jackson.module.subtype.PackageVersion;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.ServiceLoader;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+/**
+ * Subtype module.
+ * <p>
+ * The module caches the subclass, so it's non-real-time.
+ * It's for registering subtypes without annotating the parent class.
+ * See <a href="https://github.com/FasterXML/jackson-databind/issues/2104">this issues</a> in jackson-databind.
+ * <p>
+ * When not found in the cache, it loads and caches subclasses using SPI.
+ * Therefore, we can {@link #unregisterType} a class and then module will reload this class's subclasses.
+ */
+public class SubtypeModule extends Module {
+
+    private final ConcurrentHashMap<Class<?>, List<NamedType>> subtypes = new ConcurrentHashMap<>();
+
+    @Override
+    public String getModuleName() {
+        return getClass().getSimpleName();
+    }
+
+    @Override
+    public Version version() {
+        return PackageVersion.VERSION;
+    }
+
+    @Override
+    public void setupModule(SetupContext context) {
+        context.insertAnnotationIntrospector(new AnnotationIntrospector() {
+            @Override
+            public Version version() {
+                return PackageVersion.VERSION;
+            }
+
+            @Override
+            public List<NamedType> findSubtypes(Annotated a) {
+                registerTypes(a.getRawType());
+
+                List<NamedType> list1 = SubtypeModule.findSubtypes(a.getRawType(), a::getAnnotation);
+                List<NamedType> list2 = subtypes.getOrDefault(a.getRawType(), Collections.emptyList());
+
+                if (list1.isEmpty()) return list2;
+                if (list2.isEmpty()) return list1;
+                List<NamedType> list = new ArrayList<>(list1.size() + list2.size());
+                list.addAll(list1);
+                list.addAll(list2);
+                return list;
```

Comment:
Is it possible to have duplication? If so, can we have these as set first, then return as list?